File: R/dcov.R
dcov.test <-
function(x, y, index=1.0, R=NULL) {
    ## check for valid number of replicates R
    method <- "Specify the number of replicates R (R > 0) for an independence test"
    if (! is.null(R)) {
      R <- floor(R)
      if (R < 1) R <- 0
      if (R > 0) 
        method <- "dCov independence test (permutation test)"
    } else {
      R <- 0
    }

    Dx <- .arg2dist.matrix(x)
    Dy <- .arg2dist.matrix(y)
    if (!isTRUE(all.equal(index, 1.0))) {
      Dx <- Dx^index
      Dy <- Dy^index
    }
    
    n <- nrow(Dx)
    m <- nrow(Dy)
    if (n != m) stop("Sample sizes must agree")

    stat <- dcorr <- reps <- 0
    dcov <- rep(0, 4)
    if (R > 0) reps <- rep(0, R)
    pval <- 1
    dims <- c(n, ncol(Dx), ncol(Dy), R)

    # dcov = [dCov,dCor,dVar(x),dVar(y)]
    a <- .C("dCOVtest",
            x = as.double(t(Dx)),
            y = as.double(t(Dy)),
            nrow = as.integer(nrow(Dx)),
            nreps = as.integer(R),
            reps = as.double(reps),
            DCOV = as.double(dcov),
            pval = as.double(pval),
            PACKAGE = "energy")
    # test statistic is n times the square of dCov statistic
    stat <- n * a$DCOV[1]^2
    dcorr <- a$DCOV
    V <- dcorr[[1]]
    names(stat) <- "nV^2"
    names(V) <- "dCov"
    dataname <- paste("index ", index, ", replicates ", R, sep="")
    pval <- ifelse (R < 1, NA, a$pval)
    e <- list(
        statistic = stat,
        method = method,
        estimate = V,
        estimates = dcorr,
        p.value = pval,
        replicates = n* a$reps^2,
        n = n,
        data.name = dataname)
    class(e) <- "htest"
    return(e)
}

dcor.test <-
  function(x, y, index=1.0, R) {
    # distance correlation test for multivariate independence
    # like dcov.test but using dcor as the test statistic
    if (missing(R)) R <- 0
    R <- ifelse(R > 0, floor(R), 0)
    RESULT <- dcov.test(x, y, index=index, R) 
    # this test statistic is n times the square of dCov statistic
    DCOVteststat <- RESULT$statistic
    DCOVreplicates <- RESULT$replicates
    
    # RESULT$estimates = [dCov,dCor,dVar(x),dVar(y)]
    # dVar are invariant under permutation of sample indices
    estimates = RESULT$estimates
    names(estimates) <- c("dCov", "dCor", "dVar(X)", "dVar(Y)")
    
    DCORteststat <- RESULT$estimates[2]
    dvarX <- RESULT$estimates[3]
    dvarY <- RESULT$estimates[4]
    n <- RESULT$n
    
    if (R > 0) {
      DCORreps <- sqrt(DCOVreplicates / n) / sqrt(dvarX * dvarY)
      p.value <- (1 + sum(DCORreps >= DCORteststat)) / (1 + R)
    } else {
      p.value <- NA
      DCORreps <- NA
    }
    
    names(DCORteststat) <- "dCor"
    dataname <- paste("index ", index, ", replicates ", R, sep="")
    method <- ifelse(R > 0, "dCor independence test (permutation test)", 
                     "Specify the number of replicates R>0 for an independence test")
    e <- list(
      method = method,
      statistic = DCORteststat,
      estimates = estimates,
      p.value = p.value,
      replicates = DCORreps,
      n = n,
      data.name = dataname)
    class(e) <- "htest"
    return(e)
  }


.dcov <-
function(x, y, index=1.0) {
    # distance covariance statistic for independence
    # dcov = [dCov,dCor,dVar(x),dVar(y)]   (vector)
    # this function provides the fast method for computing dCov
    # it is called by the dcov and dcor functions
  
  Dx <- .arg2dist.matrix(x)
  Dy <- .arg2dist.matrix(y)
  if (!isTRUE(all.equal(index, 1.0))) {
    Dx <- Dx^index
    Dy <- Dy^index
  }
  
  n <- nrow(Dx)
  m <- nrow(Dy)
  if (n != m) stop("Sample sizes must agree")
  dims <- c(n, ncol(Dx), ncol(Dy))
    idx <- 1:dims[1]
    DCOV <- numeric(4)
    a <- .C("dCOV",
            x = as.double(t(Dx)),
            y = as.double(t(Dy)),
            nrow = as.integer(n),
            DCOV = as.double(DCOV),
            PACKAGE = "energy")
    return(a$DCOV)
}

dcov <-
function(x, y, index=1.0) {
    # distance correlation statistic for independence
    return(.dcov(x, y, index)[1])
}

dcor <-
function(x, y, index=1.0) {
    # distance correlation statistic for independence
    return(.dcov(x, y, index)[2])
}

File: src/dcov.c
/*
   dcov.c: distance correlation and covariance statistics
   and dCov test for multivariate independence

   Szekely, G.J., Rizzo, M.L., and Bakirov, N.K. (2007)
   "Measuring and testing dependence by correlation of distances"
   Annals of Statistics, Vol. 35 No. 6, pp. 2769-2794.

 Author: Maria L. Rizzo
 energy package
 github.com/mariarizzo/energy
 
 */


#include <R.h>
#include <Rmath.h>

void   dCOVtest(double *x, double *y, int *nrow, int *nreps,
                double *reps, double *DCOV, double *pval);
void   dCOV(double *x, double *y, int *nrow, double *DCOV);
double Akl(double **akl, double **A, int n);

/* functions in utilities.c */
extern double **alloc_matrix(int r, int c);
extern int    **alloc_int_matrix(int r, int c);
extern void   free_matrix(double **matrix, int r, int c);
extern void   free_int_matrix(int **matrix, int r, int c);
extern void   permute(int *J, int n);
extern void   roworder(double *x, int *byrow, int r, int c);
extern void   Euclidean_distance(double *x, double **Dx, int n, int d);
extern void   index_distance(double **Dx, int n, double index);
extern void   vector2matrix(double *x, double **y, int N, int d, int isroworder);


void dCOVtest(double *x, double *y, int *nrow, int *nreps,
              double *reps, double *DCOV, double *pval) {
    /*  input vectors must expand to distance matrices 
        any exponent must be pre-computed in R
        computes dCov(x,y), dCor(x,y), dVar(x), dVar(y)
        V-statistic is n*dCov^2 where n*dCov^2 --> Q
        DCOV  : vector [dCov, dCor, dVar(x), dVar(y), mean(A), mean(B)]
     */
    int    i, j, k, r, J, K, M;
    int    n = nrow[0], R = nreps[0];
    int*   perm;
    double **Dx, **Dy, **A, **B;
    double dcov, V;
    double n2 = (double) n * n;
    Dx = alloc_matrix(n, n);
    Dy = alloc_matrix(n, n);
  	vector2matrix(x, Dx, n, n, 1);
	  vector2matrix(y, Dy, n, n, 1);

    A = alloc_matrix(n, n);
    B = alloc_matrix(n, n);
    Akl(Dx, A, n);
    Akl(Dy, B, n);
    free_matrix(Dx, n, n);
    free_matrix(Dy, n, n);

    /* compute dCov(x,y), dVar(x), dVar(y) */
    for (k=0; k<4; k++)
        DCOV[k] = 0.0;
    for (k=0; k<n; k++)
        for (j=0; j<n; j++) {
            DCOV[0] += A[k][j]*B[k][j];
            DCOV[2] += A[k][j]*A[k][j];
            DCOV[3] += B[k][j]*B[k][j];
        }

    for (k=0; k<4; k++) {
        DCOV[k] /= n2;
        if (DCOV[k] > 0)
            DCOV[k] = sqrt(DCOV[k]);
            else DCOV[k] = 0.0;
    }
    /* compute dCor(x, y) */
    V = DCOV[2]*DCOV[3];
    if (V > DBL_EPSILON)
        DCOV[1] = DCOV[0] / sqrt(V);
        else DCOV[1] = 0.0;

	if (R > 0) {
		/* compute the replicates */
		if (DCOV[1] > 0.0) {
			perm = Calloc(n, int);
			M = 0;
			for (i=0; i<n; i++) perm[i] = i;
      GetRNGstate();
      for (r=0; r<R; r++) {
			   permute(perm, n);
			   dcov = 0.0;
			   for (k=0; k<n; k++) {
				   K = perm[k];
				   for (j=0; j<n; j++) {
					   J = perm[j];
					   dcov += A[k][j]*B[K][J];
				   }
			   }
			   dcov /= n2;
			   dcov = sqrt(dcov);
			   reps[r] = dcov;
			   if (dcov >= DCOV[0]) M++;
			}
			*pval = (double) (M+1) / (double) (R+1);
      PutRNGstate();
			Free(perm);
		} else {
		    *pval = 1.0;
			}
	}

    free_matrix(A, n, n);
    free_matrix(B, n, n);
    return;
}

void dCOV(double *x, double *y, int *nrow, double *DCOV) {
  /*  input vectors must expand to distance matrices 
      any exponent must be pre-computed in R
        computes dCov(x,y), dCor(x,y), dVar(x), dVar(y)
        V-statistic is n*dCov^2 where n*dCov^2 --> Q
        DCOV  : vector [dCov, dCor, dVar(x), dVar(y)]
     */

    int    j, k, n = nrow[0];
    double **Dx, **Dy, **A, **B;
    double V, n2 = (double) n * n;

    Dx = alloc_matrix(n, n);
    Dy = alloc_matrix(n, n);
		vector2matrix(x, Dx, n, n, 1);
		vector2matrix(y, Dy, n, n, 1);

    A = alloc_matrix(n, n);
    B = alloc_matrix(n, n);
    Akl(Dx, A, n);
    Akl(Dy, B, n);
    free_matrix(Dx, n, n);
    free_matrix(Dy, n, n);

    n2 = ((double) n) * n;

    /* compute dCov(x,y), dVar(x), dVar(y) */
    for (k=0; k<4; k++)
        DCOV[k] = 0.0;
    for (k=0; k<n; k++)
        for (j=0; j<n; j++) {
            DCOV[0] += A[k][j]*B[k][j];
            DCOV[2] += A[k][j]*A[k][j];
            DCOV[3] += B[k][j]*B[k][j];
        }

    for (k=0; k<4; k++) {
        DCOV[k] /= n2;
        if (DCOV[k] > 0)
            DCOV[k] = sqrt(DCOV[k]);
            else DCOV[k] = 0.0;
    }
    /* compute dCor(x, y) */
    V = DCOV[2]*DCOV[3];
    if (V > DBL_EPSILON)
        DCOV[1] = DCOV[0] / sqrt(V);
        else DCOV[1] = 0.0;

    free_matrix(A, n, n);
    free_matrix(B, n, n);
    return;
}

double Akl(double **akl, double **A, int n) {
    /* -computes the A_{kl} or B_{kl} distances from the
        distance matrix (a_{kl}) or (b_{kl}) for dCov, dCor, dVar
        dCov = mean(Akl*Bkl), dVar(X) = mean(Akl^2), etc.
    */
    int j, k;
    double *akbar;
    double abar;

    akbar = Calloc(n, double);
    abar = 0.0;
    for (k=0; k<n; k++) {
        akbar[k] = 0.0;
        for (j=0; j<n; j++) {
            akbar[k] += akl[k][j];
        }
        abar += akbar[k];
        akbar[k] /= (double) n;
    }
    abar /= (double) (n*n);

    for (k=0; k<n; k++)
        for (j=k; j<n; j++) {
            A[k][j] = akl[k][j] - akbar[k] - akbar[j] + abar;
            A[j][k] = A[k][j];
        }
    Free(akbar);
    return(abar);
}
